File: src/main/java/rag/app/stages/RerankingStage.java
package rag.app.stages;

import rag.app.Context;
import rag.app.StrategyRegistry;
import rag.rerank.Reranker;
import rag.retrieval.Hit;
import rag.trace.TraceBus;
import rag.trace.TraceEvent;

import java.util.List;

public class RerankingStage implements PipelineStage {

    private final Reranker reranker;

    public RerankingStage(StrategyRegistry registry) {
        this.reranker = registry.getReranker();
    }

    @Override
    public String getName() {
        return "RerankingStage";
    }

    @Override
    public void run(Context context, TraceBus traceBus) throws Exception {

        long start = System.currentTimeMillis();
        String error = null;
        List<Hit> reranked = null;
        Exception failure = null;

        try {
            reranked = reranker.rerank(
                    context.getTerms(),
                    context.getHits(),
                    null
            );
            context.setHits(reranked);

        } catch (Exception ex) {
            error = ex.getMessage();
            failure = ex;
        } finally {
            long duration = System.currentTimeMillis() - start;

            traceBus.publish(
                    new TraceEvent(
                            getName(),
                            "top=" + (reranked != null && !reranked.isEmpty()
                                    ? reranked.get(0)
                                    : "none"),
                            duration,
                            error
                    )
            );
        }
        if (failure != null) {
            throw failure;
        }
    }
}

File: src/main/python/rag/app/stages/reranking_stage.py
from typing import List, Optional
from rag.app.stages.pipeline_stage import PipelineStage
from rag.app.context import Context
from rag.app.strategy_registry import StrategyRegistry
from rag.rerank.reranker import Reranker
from rag.retrieval.hit import Hit
from rag.trace.trace_bus import TraceBus
from rag.trace.trace_event import TraceEvent
from rag.utils.timer_utils import now


class RerankingStage(PipelineStage):
    def __init__(self, registry: StrategyRegistry) -> None:
        self.reranker: Reranker = registry.get_reranker()

    def get_name(self) -> str:
        return "RerankingStage"

    def run(self, context: Context, trace_bus: TraceBus) -> None:
        start: int = now()
        error: Optional[str] = None
        reranked: Optional[List[Hit]] = None
        failure: Optional[Exception] = None

        try:
            reranked = self.reranker.rerank(
                context.get_terms() or [],
                context.get_hits() or [],
                {"query": context.get_question()}
            )
            context.set_hits(reranked)
        except Exception as ex:
            error = str(ex)
            failure = ex
        finally:
            duration: int = now() - start
            top_hit: str = str(reranked[0]) if reranked and reranked else "none"
            trace_bus.publish(
                TraceEvent(
                    self.get_name(),
                    f"top={top_hit}",
                    duration,
                    error
                )
            )

        if failure is not None:
            raise failure






